### ISSUE TITLE
Can't run swarm tests in Docker for Mac

### ISSUE THREAD

**Issue by aaronlehmann** [docker/docker#23545]
I tried to run `DockerSwarmSuite.TestApiSwarmPromoteDemote` using Docker for Mac as the container runtime.

The test fails with this output:

```
FAIL: docker_api_swarm_test.go:161: DockerSwarmSuite.TestApiSwarmPromoteDemote

[d56553497] waiting for daemon to start
docker_api_swarm_test.go:162:
[d56553497] exiting daemon
    d1 := s.AddDaemon(c, false, false)
check_test.go:212:
    c.Assert(err, check.IsNil)
... value *errors.errorString = &errors.errorString{s:"[d56553497] Daemon exited during startup"} ("[d56553497] Daemon exited during startup")

OOPS: 0 passed, 1 FAILED
--- FAIL: Test (0.38s)
FAIL
coverage: 12.9% of statements
```

Inspecting the test logs, I found an error binding to a unix socket when starting a daemon:

```
time="2016-06-14T21:00:02.103033648Z" level=fatal msg="can't create unix socket /go/src/github.com/docker/docker/bundles/1.12.0-dev/test-integration-cli/d86922137/docker.sock: listen unix /go/src/github.com/docker/docker/bundles/1.12.0-dev/test-integration-cli/d86922137/docker.sock: bind: file name too long"
```

The filename length limit for a unix socket should be 108 characters, but this path is only 95 characters, so I am puzzled why `bind` is failing.

I seem to have the same problem with all other tests in `DockerSwarmSuite`.

**Comment by icecrime** [comment]
Ping @justincormack!

**Comment by AkihiroSuda** [comment]
Perhaps the limit is 92 in your system? Although I thought the limit is 104 chars for OSX...

http://pubs.opengroup.org/onlinepubs/009604499/basedefs/sys/un.h.html
> The size of sun_path has intentionally been left undefined. This is because different implementations use different sizes. For example, 4.3 BSD uses a size of 108, and 4.4 BSD uses a size of 104. Since most implementations originate from BSD versions, the size is typically in the range 92 to 108.

**Comment by aaronlehmann** [comment]
I evaluated `sizeof(sockaddr_un.sun_path)` inside the container and it was 108. It's possible that for some reason, the actual limit is lower. However, it's strange that these tests appear to work in CI and on other peoples' Linux systems.

**Comment by justincormack** [comment]
There are issues with socket name lengths if the socket is being created on a volume shared with the Mac, as the path being created will be the full un-namespaced socket path _and_ will also include the OSX directory path you are sharing, _and_ the OSX length is shorter than the Linux one. This also happens with running the docker/docker tests, unless you run them inside a container. We do have an outstanding issue for it, but it will probably have to be a really hacky workaround (eg remapping paths to something else) and it will probably not happen for a while. These socket name lengths are a relic of an older more innocent age when filesystem path lengths were smaller, but are stuck in the ABI. I might try to change the tests to use sockets in `/tmp` or something as a workaround...

**Comment by cpuguy83** [comment]
Yes I always run the test suite without setting BINDDIR, which also means having to set `DOCKER_HOST=unix:///var/run/docker.sock` because the Makefile will automatically bind the bundles dir if it is not set.

@aaronlehmann Can you test to make sure you aren't setting a binddir and you have `DOCKER_HOST` set as above?

**Comment by aaronlehmann** [comment]
The strange thing is that I don't think this socket is on a shared volume. Unless something has changed recently, this build environment is copied to the container rather than mounted. I will take a closer look to make sure there's no volume behind this.

**Comment by aaronlehmann** [comment]
> Yes I always run the test suite without setting BINDDIR, which also means having to set DOCKER_HOST=unix:///var/run/docker.sock because the Makefile will automatically bind the bundles dir if it is not set.
> @aaronlehmann Can you test to make sure you aren't setting a binddir and you have DOCKER_HOST set as above?
> 

Aha, will give this a try.

Is there anything we can do to improve the defaults? It would be nice if people could do dev with Docker for Mac without resorting to workarounds.

**Comment by aaronlehmann** [comment]
@cpuguy83: Setting `BINDDIR=` and `DOCKER_HOST=unix:///var/run/docker.sock` solved the problem.

Should I close this ticket? I still think it would be nice if integration tests worked out of the box without special settings.

**Comment by tkopczynski** [comment]
I had the same problem but managed to work around this by creating a shorter path on my host ;) 
Anyway, I think this should be mentioned somewhere in the docs. I can create a PR for this if you agree that it it's ok to put a workaround in the documentation and a proper solution can't be achieved at the moment.

**Comment by justincormack** [comment]
We definitely want to fix this, and we should document it now, not sure where it should be - the docker for mac docs will be moving into the main documentation soon I think so maybe then.

**Comment by stevvooe** [comment]
Can we just put these sockets elsewhere? Say, in temp?

I'm see this problem in the swarm tests that creating a bunch of subdaemons.

**Comment by cpuguy83** [comment]
#25171 to move these to a temp dir.

**Comment by justincormack** [comment]
@aaronlehmann this should be fixed now - can you check?

**Comment by cpuguy83** [comment]
It's not :(

**Comment by cpuguy83** [comment]
Actually, I found my issue, related to setting DOCKER_GRAPHDRIVER, which is unrelated to d4m.

I think we can close this.
@aaronlehmann Have you had any other issues?

**Comment by aaronlehmann** [comment]
When I remove the `BINDDIR= DOCKER_HOST=unix:///var/run/docker.sock` environment variable settings, tests fail like this (different error than before):

```
FAIL: docker_api_service_update_test.go:20: DockerSwarmSuite.TestApiServiceUpdatePort

[d5215bb978a57] waiting for daemon to start
[d5215bb978a57] daemon started
docker_api_service_update_test.go:21:
    d := s.AddDaemon(c, true, true)
check_test.go:268:
    c.Assert(err, check.IsNil)
... value *errors.errorString = &errors.errorString{s:"could not load busybox image: operation not supported\n"} ("could not load busybox image: operation not supported\n")
```

**Comment by justincormack** [comment]
@aaronlehmann can you `strace` that operation not supported and see what it is? It may be a missing osxfs feature which maybe we can prioritise.

**Comment by aaronlehmann** [comment]
@justincormack: I think it's related to xattrs:

```
[pid  8010] lgetxattr("/go/src/github.com/docker/docker/bundles/1.13.0-dev/test-integration-cli/d1478b941277c/root/overlay/2072f8d99635d3f07afa6401a47dba404cc79540c908398215261f8229ff227a/root", "security.capability", 0xc420388580, 128) = -1 EOPNOTSUPP (Operation not supported)
```

**Comment by cpuguy83** [comment]
Yeah, looks like you are running tests with the overlay driver, which for swarm and daemon tests is going to run on osxfs if you have mounted that in.

**Comment by aaronlehmann** [comment]
Ah, I seem to have `export DOCKER_GRAPHDRIVER="overlay"` in my environment. Let me try with aufs.

**Comment by cpuguy83** [comment]
I don't think any graphdriver other than vfs would work on top of osxfs.

**Comment by aaronlehmann** [comment]
I have other issues trying to run the tests on aufs on top of osxfs:

```
FAIL: docker_api_service_update_test.go:20: DockerSwarmSuite.TestApiServiceUpdatePort

[de5ed7de02e9e] waiting for daemon to start
[de5ed7de02e9e] daemon started
docker_api_service_update_test.go:26:
    waitAndAssert(c, defaultReconciliationTimeout, d.checkActiveContainerCount, checker.Equals, 1)
docker_utils.go:1535:
    c.Assert(v, checker, args...)
... obtained int = 0
... expected int = 1

[de5ed7de02e9e] exiting daemon
```

It seems to fail later on. Some tests actually pass.

**Comment by zalenet** [comment]
我正好也遇到了这个问题，情况是我在mac 上使用docker 部署mysql 出现了一模一样的问题
![image](https://user-images.githubusercontent.com/46991944/97290396-7523e080-1883-11eb-921f-b7975333a0b7.png)
启动mysql 后两到5秒钟就处于Restarting 状态了
以下为log 
![image](https://user-images.githubusercontent.com/46991944/97290549-a4d2e880-1883-11eb-97b7-bc047630660d.png)
我不知道是否是这个原因，但是我通过这个方法解决了
mysql 数据挂载是当前目录的data 文件夹中 具体路径为
![image](https://user-images.githubusercontent.com/46991944/97290717-e8c5ed80-1883-11eb-91a1-3e4d0e6b9300.png)
可以看到release 开头的文件夹名称特别长，当我将文件夹名称改短后，msyql 启动就正常了
![image](https://user-images.githubusercontent.com/46991944/97290940-35a9c400-1884-11eb-961b-3ce5393251e9.png)
